vim-patch: 9.1.{1341,1344}

glepnir opened the pull request:
vim-patch:9.1.1341: cannot define completion triggers

Problem:  Cannot define completion triggers and act upon it
Solution: add the new option 'isexpand' and add the complete_match()
          function to return the completion matches according to the
          'isexpand' setting (glepnir)

Currently, completion trigger position is determined solely by the
'iskeyword' pattern (\k\+$), which causes issues when users need
different completion behaviors - such as triggering after '/' for
comments or '.' for methods. Modifying 'iskeyword' to include these
characters has undesirable side effects on other Vim functionality that
relies on keyword definitions.

Introduce a new buffer-local option 'isexpand' that allows specifying
different completion triggers and add the complete_match() function that
finds the appropriate start column for completion based on these
triggers, scanning backwards from cursor position.

This separation of concerns allows customized completion behavior
without affecting iskeyword-dependent features. The option's
buffer-local nature enables per-filetype completion triggers.

closes: https://github.com/vim/vim/pull/16716

https://github.com/vim/vim/commit/bcd5995b40a1c26e735bc326feb2e3ac4b05426b

Co-authored-by: glepnir <glephunter@gmail.com>

patch 9.1.1344: double free in f_complete_match() (after v9.1.1341)

Problem:  double free in f_complete_match() (after v9.1.1341)
Solution: remove additional free of trig pointer, correctly free
          regmatch.regprog and before_cursor in the error case

closes: https://github.com/vim/vim/pull/17203

https://github.com/vim/vim/commit/3accf046ec3d0ee4a762d15452ae46596e1a0540

Co-authored-by: Christian Brabandt <cb@256bit.org>

glepnir committed:
vim-patch:9.1.1341: cannot define completion triggers

Problem:  Cannot define completion triggers and act upon it
Solution: add the new option 'isexpand' and add the complete_match()
          function to return the completion matches according to the
          'isexpand' setting (glepnir)

Currently, completion trigger position is determined solely by the
'iskeyword' pattern (\k\+$), which causes issues when users need
different completion behaviors - such as triggering after '/' for
comments or '.' for methods. Modifying 'iskeyword' to include these
characters has undesirable side effects on other Vim functionality that
relies on keyword definitions.

Introduce a new buffer-local option 'isexpand' that allows specifying
different completion triggers and add the complete_match() function that
finds the appropriate start column for completion based on these
triggers, scanning backwards from cursor position.

This separation of concerns allows customized completion behavior
without affecting iskeyword-dependent features. The option's
buffer-local nature enables per-filetype completion triggers.

closes: vim/vim#16716

https://github.com/vim/vim/commit/bcd5995b40a1c26e735bc326feb2e3ac4b05426b

Co-authored-by: glepnir <glephunter@gmail.com>

glepnir committed:
vim-patch:9.1.1344: double free in f_complete_match() (after v9.1.1341)

Problem:  double free in f_complete_match() (after v9.1.1341)
Solution: remove additional free of trig pointer, correctly free
          regmatch.regprog and before_cursor in the error case

closes: https://github.com/vim/vim/pull/17203

https://github.com/vim/vim/commit/3accf046ec3d0ee4a762d15452ae46596e1a0540

Co-authored-by: Christian Brabandt <cb@256bit.org>

justinmk commented on src/nvim/eval.lua:
```diff
@@ -1477,6 +1477,59 @@ M.funcs = {
+    args = { 0, 2 },
+    base = 0,
+    desc = [=[
+      Returns a List of matches found according to the 'isexpand'
```
matches found *where* ? in the buffer text and cursor position?

glepnir replied:
```
      When no arguments are provided, uses the current cursor
      position.
```

glepnir committed:
vim-patch:ffc89e4 runtime(doc): clarify complete_match() and 'isexpand' option

clarify complete_match() documentation to better explain its backward
search behavior, argument handling, and return value format and add an
example of isexpand

closes: https://github.com/vim/vim/pull/17212

https://github.com/vim/vim/commit/ffc89e47d014178bcd0a681ff2c8e18470cc972b

zeertzjq commented on src/nvim/eval.lua:
```diff
@@ -1477,6 +1477,59 @@ M.funcs = {
+
+      Examples: >
+        set isexpand=.,->,/,/*,abc
+        func CustomComplete()
+        let res = complete_match()
+        if res->len() == 0 | return | endif
+        let [col, trigger] = res[0]
+        let items = []
+        if trigger == '/*'
+          let items = ['/** */']
+        elseif trigger == '/'
+          let items = ['/*! */', '// TODO:', '// fixme:']
+        elseif trigger == '.'
+          let items = ['length()']
+        elseif trigger =~ '^\->'
+          let items = ['map()', 'reduce()']
+        elseif trigger =~ '^\abc'
+          let items = ['def', 'ghk']
+        endif
+        if items->len() > 0
+          let startcol = trigger =~ '^/' ? col : col + len(trigger)
+          call complete(startcol, items)
+        endif
+        endfunc
```
Function body is not indented

zeertzjq commented on src/nvim/eval.lua:
```diff
@@ -1477,6 +1477,59 @@ M.funcs = {
+        endfunc
+        inoremap <Tab> <Cmd>call CustomComplete()<CR>
+<
+
+      Return type: list<list<any>>
```
Remove this. This is Vim9 script notation.

zeertzjq commented on src/nvim/eval.lua:
```diff
@@ -1477,6 +1477,59 @@ M.funcs = {
+      Return type: list<list<any>>
+    ]=],
+    name = 'complete_match',
+    params = { { 'lnum', 'integer?' }, { 'col', 'integer?' } },
```
You don't need to add `?` here.

zeertzjq commented on src/nvim/insexpand.c:
```diff
@@ -3081,6 +3081,103 @@ void f_complete_check(typval_T *argvars, typval_T *rettv, EvalFuncData fptr)
 }
 
+/// Add match item to the return list.
+/// Returns FAIL if out of memory, OK otherwise.
+static int add_match_to_list(typval_T *rettv, char *str, int pos)
```
```suggestion
static void add_match_to_list(typval_T *rettv, char *str, int pos)
```

zeertzjq commented on src/nvim/insexpand.c:
```diff
@@ -3081,6 +3081,103 @@ void f_complete_check(typval_T *argvars, typval_T *rettv, EvalFuncData fptr)
+static int add_match_to_list(typval_T *rettv, char *str, int pos)
+{
+  list_T *match = tv_list_alloc(kListLenMayKnow);
+  if (match == NULL) {
+    return FAIL;
+  }
+
```
```suggestion
```

zeertzjq commented on src/nvim/insexpand.c:
```diff
@@ -3081,6 +3081,103 @@ void f_complete_check(typval_T *argvars, typval_T *rettv, EvalFuncData fptr)
+  tv_list_append_number(match, pos + 1);
+  tv_list_append_string(match, str, -1);
+  tv_list_append_list(rettv->vval.v_list, match);
+  return OK;
```
```suggestion
```

zeertzjq commented on src/nvim/insexpand.c:
```diff
@@ -3081,6 +3081,103 @@ void f_complete_check(typval_T *argvars, typval_T *rettv, EvalFuncData fptr)
+        }
+
+        int bytepos = (int)(regmatch.startp[0] - before_cursor);
+        int ret = add_match_to_list(rettv, trig, bytepos);
```
```suggestion
        add_match_to_list(rettv, trig, bytepos);
```

zeertzjq commented on src/nvim/insexpand.c:
```diff
@@ -3081,6 +3081,103 @@ void f_complete_check(typval_T *argvars, typval_T *rettv, EvalFuncData fptr)
+        int bytepos = (int)(regmatch.startp[0] - before_cursor);
+        int ret = add_match_to_list(rettv, trig, bytepos);
+        xfree(trig);
+        if (ret == FAIL) {
+          xfree(before_cursor);
+          vim_regfree(regmatch.regprog);
+          return;
+        }
```
```suggestion
```

zeertzjq commented on src/nvim/insexpand.c:
```diff
@@ -3081,6 +3081,103 @@ void f_complete_check(typval_T *argvars, typval_T *rettv, EvalFuncData fptr)
+      if (len > 0 && (int)len <= col) {
+        if (strncmp(cur_end - len, part, len) == 0) {
+          int bytepos = col - (int)len;
+          if (add_match_to_list(rettv, part, bytepos) == FAIL) {
+            xfree(before_cursor);
+            return;
+          }
```
```suggestion
          add_match_to_list(rettv, part, bytepos);
```

zeertzjq commented on src/nvim/insexpand.c:
```diff
@@ -3081,6 +3081,103 @@ void f_complete_check(typval_T *argvars, typval_T *rettv, EvalFuncData fptr)
+  }
+
+  char *before_cursor = xstrnsave(line, (size_t)col);
+  if (before_cursor == NULL) {
+    return;
+  }
```
```suggestion
```

zeertzjq commented on src/nvim/insexpand.c:
```diff
@@ -3081,6 +3081,86 @@ void f_complete_check(typval_T *argvars, typval_T *rettv, EvalFuncData fptr)
+    return;
+  }
+
+  char *before_cursor = xstrnsave(line, (size_t)col);
```
I think you should keep the blank line after this line.

zeertzjq commented on src/nvim/insexpand.c:
```diff
@@ -3081,6 +3081,86 @@ void f_complete_check(typval_T *argvars, typval_T *rettv, EvalFuncData fptr)
+  }
+
+  char *before_cursor = xstrnsave(line, (size_t)col);
+  if (ise == NULL || *ise == NUL) {
```
I think you should keep the blank line above this line.

glepnir replied:
okay kept

zeertzjq commented on src/nvim/insexpand.c:
```diff
@@ -3081,6 +3081,86 @@ void f_complete_check(typval_T *argvars, typval_T *rettv, EvalFuncData fptr)
+    if (regmatch.regprog != NULL) {
+      if (vim_regexec_nl(&regmatch, before_cursor, (colnr_T)0)) {
+        char *trig = xstrnsave(regmatch.startp[0], (size_t)(regmatch.endp[0] - regmatch.startp[0]));
+        if (trig == NULL) {
+          xfree(before_cursor);
+          return;
+        }
+
```
```suggestion
```

zeertzjq commented:
Missing commit URL in the last commit

zeertzjq commented:
Missing change to usr_41.txt

zeertzjq commented on src/nvim/insexpand.c:
```diff
@@ -3081,6 +3081,86 @@ void f_complete_check(typval_T *argvars, typval_T *rettv, EvalFuncData fptr)
+/// Add match item to the return list.
+static void add_match_to_list(typval_T *rettv, char *str, int pos)
+{
+  list_T *match = tv_list_alloc(kListLenMayKnow);
```
```suggestion
  list_T *match = tv_list_alloc(2);
```

zeertzjq commented on src/nvim/insexpand.c:
```diff
@@ -3081,6 +3081,86 @@ void f_complete_check(typval_T *argvars, typval_T *rettv, EvalFuncData fptr)
+/// "complete_match()" function
+void f_complete_match(typval_T *argvars, typval_T *rettv, EvalFuncData fptr)
+{
+  tv_list_alloc_ret(rettv,   kListLenUnknown);
```
```suggestion
  tv_list_alloc_ret(rettv, kListLenUnknown);
```

zeertzjq commented on src/nvim/option.c:
```diff
@@ -4460,6 +4460,8 @@ void *get_varp_scope_from(vimoption_T *p, int opt_flags, buf_T *buf, win_T *win)
       return &(buf->b_p_def);
     case kOptInclude:
       return &(buf->b_p_inc);
+    case kOptIsexpand:
+      return &(buf->b_p_ise);
     case kOptCompleteopt:
       return &(buf->b_p_cot);
```
```suggestion
    case kOptCompleteopt:
      return &(buf->b_p_cot);
    case kOptIsexpand:
      return &(buf->b_p_ise);
```

zeertzjq commented on src/nvim/option.c:
```diff
@@ -4545,6 +4547,8 @@ void *get_varp_from(vimoption_T *p, buf_T *buf, win_T *win)
     return *buf->b_p_def != NUL ? &(buf->b_p_def) : p->var;
   case kOptInclude:
     return *buf->b_p_inc != NUL ? &(buf->b_p_inc) : p->var;
+  case kOptIsexpand:
+    return *buf->b_p_ise != NUL ? &(buf->b_p_ise) : p->var;
   case kOptCompleteopt:
     return *buf->b_p_cot != NUL ? &(buf->b_p_cot) : p->var;
```
```suggestion
  case kOptCompleteopt:
    return *buf->b_p_cot != NUL ? &(buf->b_p_cot) : p->var;
  case kOptIsexpand:
    return *buf->b_p_ise != NUL ? &(buf->b_p_ise) : p->var;
```

zeertzjq commented on src/nvim/option_vars.h:
```diff
@@ -376,6 +376,7 @@ EXTERN char *p_indk;            ///< 'indentkeys'
 EXTERN char *p_icm;             ///< 'inccommand'
 EXTERN char *p_isf;             ///< 'isfname'
 EXTERN char *p_isi;             ///< 'isident'
+EXTERN char *p_ise;             ///< 'isexpand'
```
```suggestion
EXTERN char *p_ise;             ///< 'isexpand'
EXTERN char *p_isf;             ///< 'isfname'
EXTERN char *p_isi;             ///< 'isident'
```

zeertzjq commented on src/nvim/buffer_defs.h:
```diff
@@ -558,6 +558,7 @@ struct file_buffer {
   char *b_p_fo;                 ///< 'formatoptions'
   char *b_p_flp;                ///< 'formatlistpat'
   int b_p_inf;                  ///< 'infercase'
+  char *b_p_ise;                ///< 'isexpand'
```
```suggestion
  char *b_p_ise;                ///< 'isexpand' local value
```

zeertzjq commented:
- Missing original commit URL in the last commit
- Missing changes to usr_41.txt and optwin.vim
- Missing change to `buf_copy_options()`
- Missing change to gen_opt_test.vim

zeertzjq commented on runtime/optwin.vim:
```diff
@@ -1102,6 +1102,8 @@ call <SID>AddOption("isfname", gettext("specifies the characters in a file name"
 call <SID>AddOption("isident", gettext("specifies the characters in an identifier"))
 call <SID>OptionG("isi", &isi)
+call <SID>AddOption("isexpand", gettext("defines trigger strings for complete_match()"))
+ call append("$", "\t" .. s:local_to_buffer)
```
```suggestion
call append("$", "\t" .. s:local_to_buffer)
```

zeertzjq commented:
Still missing change to `buf_copy_options()`

glepnir commented:
oops added

zeertzjq commented on runtime/doc/usr_41.txt:
```diff
@@ -934,6 +934,7 @@ Insert mode completion:				*completion-functions*
 	complete_add()		add to found matches
 	complete_check()	check if completion should be aborted
 	complete_info()		get current completion information
+	complete_match()	get completion column and trigger text
```
This doesn't match the description in Vim

zeertzjq commented on src/nvim/options.lua:
```diff
@@ -4628,6 +4628,33 @@ local options = {
+      abbreviation = 'ise',
+      cb = 'did_set_isexpand',
+      defaults = '',
+      deny_duplicates = false,
```
```suggestion
      deny_duplicates = true,
```
